File: lessons/00-workflow-or-agent/typescript/src/fake-decision-provider.ts
import {readFileSync} from "node:fs";

import {parse} from "yaml";

import {FAKE_MODEL_PATH} from "./paths.js";
import type {DecisionProvider} from "./types.js";

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

export class ScriptedDecisionProvider implements DecisionProvider {
  readonly #responses: Readonly<Record<string, unknown>>;

  constructor(responses: Record<string, unknown>) {
    this.#responses = structuredClone(responses);
  }

  static fromFile(path: string = FAKE_MODEL_PATH): ScriptedDecisionProvider {
    const document = parse(readFileSync(path, "utf8"));
    if (!isRecord(document) || document.schema_version !== "1.0" || document.scenario_id !== "support-triage") {
      throw new Error("Unsupported scripted decision fixture.");
    }
    if (!isRecord(document.responses)) {
      throw new Error("Scripted decision fixture must contain a responses object.");
    }

    const responses: Record<string, unknown> = {};
    for (const [caseId, entry] of Object.entries(document.responses)) {
      if (!isRecord(entry) || !("return" in entry)) {
        throw new Error(`Scripted response ${caseId} must contain a return value.`);
      }
      responses[caseId] = structuredClone(entry.return);
    }
    return new ScriptedDecisionProvider(responses);
  }

  async classify(input: {caseId: string; request: string}): Promise<unknown> {
    void input.request;
    if (!(input.caseId in this.#responses)) {
      throw new Error(`Missing scripted decision for case: ${input.caseId}`);
    }
    return structuredClone(this.#responses[input.caseId]);
  }
}

File: lessons/00-workflow-or-agent/python/src/boundrelay_m0/fake_decision_provider.py
from copy import deepcopy
from pathlib import Path
from typing import Mapping

import yaml

from .paths import FAKE_MODEL_PATH


class ScriptedDecisionProvider:
    def __init__(self, responses: Mapping[str, object]) -> None:
        self._responses = deepcopy(dict(responses))

    @classmethod
    def from_file(cls, path: Path = FAKE_MODEL_PATH) -> "ScriptedDecisionProvider":
        document = yaml.safe_load(path.read_text(encoding="utf-8"))
        if (
            not isinstance(document, Mapping)
            or document.get("schema_version") != "1.0"
            or document.get("scenario_id") != "support-triage"
        ):
            raise ValueError("Unsupported scripted decision fixture.")
        entries = document.get("responses")
        if not isinstance(entries, Mapping):
            raise ValueError("Scripted decision fixture must contain a responses object.")
        responses: dict[str, object] = {}
        for case_id, entry in entries.items():
            if not isinstance(case_id, str) or not isinstance(entry, Mapping) or "return" not in entry:
                raise ValueError(f"Scripted response {case_id} must contain a return value.")
            responses[case_id] = deepcopy(entry["return"])
        return cls(responses)

    def classify(self, *, case_id: str, request: str) -> object:
        del request
        if case_id not in self._responses:
            raise ValueError(f"Missing scripted decision for case: {case_id}")
        return deepcopy(self._responses[case_id])
